concordia/views.py

```diff
@@ -209,25 +211,55 @@ def registration_rate(self, group, request):
 @method_decorator(never_cache, name="dispatch")
 class ConcordiaRegistrationView(RatelimitMixin, RegistrationView):
     form_class = UserRegistrationForm
+    ratelimit_group = "registration"
     ratelimit_key = "ip"
     ratelimit_rate = registration_rate
     ratelimit_method = "POST"
-    ratelimit_block = True
+    ratelimit_block = settings.RATELIMIT_BLOCK
 
 
 @method_decorator(never_cache, name="dispatch")
 class ConcordiaLoginView(RatelimitMixin, LoginView):
-    ratelimit_key = "ip"
+    ratelimit_group = "login"
+    ratelimit_key = "post:username"
     ratelimit_rate = "3/15m"
     ratelimit_method = "POST"
-    ratelimit_block = True
+    ratelimit_block = False
+
+    def post(self, request, *args, **kwargs):
+        blocked = is_ratelimited(
+            request,
+            group=self.ratelimit_group,
+            key=self.ratelimit_key,
+            method=self.ratelimit_method,
+            rate=self.ratelimit_rate,
+        )
+        recent_captcha = (
+            time.time() - request.session.get("captcha_validation_time", 0) < 86400
+        )
+
+        if blocked and not recent_captcha:
+            raise Ratelimited()
+        else:
+            return super().post(request, *args, **kwargs)
 
 
 def ratelimit_view(request, exception=None):
-    template_name = "429.html"
     status_code = 429
+
+    # TODO: return JSON response for JSON requests
```

Comment:
I'm thinking this is as simple as building the context dictionary below and then doing the Accept check to return either `JsonResponse` or `HttpResponse` as below

Reply:
09a4009a52491a1eba0caf5c9b4ae4f3affe8d49 implements this so I'm marking this as resolved.